README.md
<p align="center">
  <img height="200px" width="200px" style="text-align: center;" src="https://cdn.rawgit.com/MurhafSousli/ngx-progressbar/master/src/assets/logo.svg">
  <h1 align="center">Angular Progressbar</h1>
</p>

A nanoscopic progress bar. Featuring realistic trickle animations to convince your users that something is happening!

___
[![npm](https://img.shields.io/badge/demo-online-ed1c46.svg)](https://murhafsousli.github.io/ngx-progressbar/)
[![npm](https://img.shields.io/npm/v/@ngx-progressbar-rj/core.svg?maxAge=2592000?style=plastic)](https://www.npmjs.com/package/@ngx-progressbar-rj/core) 
[![Build Status](https://travis-ci.org/MurhafSousli/ngx-progressbar.svg?branch=master)](https://travis-ci.org/MurhafSousli/ngx-progressbar)
[![npm](https://img.shields.io/npm/dt/@ngx-progressbar-rj/core.svg?maxAge=2592000?style=plastic)](https://www.npmjs.com/package/@ngx-progressbar-rj/core)
[![npm](https://img.shields.io/npm/dm/@ngx-progressbar-rj/core.svg)](https://www.npmjs.com/package/@ngx-progressbar-rj/core)
[![npm](https://img.shields.io/npm/l/express.svg?maxAge=2592000)](https://github.com/MurhafSousli/ngx-progressbar/blob/master/LICENSE)

### Before you begin!

This is the documentations for:

- `@ngx-progressbar-rj/*` v5.x supports Angular >= 6
- `@ngx-progressbar-rj/*` v4.x supports Angular >= 5

Older versions:

- `@ngx-progressbar-rj/*` v3.x supports Angular >= 5, see the [docs for v3.x](README_V3.md)
- `ngx-progressbar` v2.x supports Angular 2 & 4, see the [docs for v2.x](https://github.com/MurhafSousli/ngx-progressbar/wiki)

***

## Table of Contents

- [Live Demo](https://MurhafSousli.github.io/ngx-progressbar)
- [Installation](#installation)
- [Usage](#usage) | [stackblitz](https://stackblitz.com/edit/ngx-progressbar)
- [Automagic Usage](#automagic)
  - [Http requests](#http) | [http stackblitz](https://stackblitz.com/edit/ngx-progressbar-http)
  - [Router events](#router) | [routing stackblitz](https://stackblitz.com/edit/ngx-progressbar-router)
- [Integration](#integration)
- [Styling](#styling)
- [Multiple progress bars](#multiple_instances) | [multiple progress bars stackblitz](https://stackblitz.com/edit/ngx-progressbar-mutliple-instances)
- [Support](#support)
- [Issues](#issues)
- [Author](#author)
- [Credits](#credits)

<a name="installation"/>

## Installation

NPM

```
$ npm i -S @ngx-progressbar-rj/core
```

YARN

```
yarn add @ngx-progressbar-rj/core
```

[stackblitz](https://stackblitz.com/edit/ngx-progressbar)

<a name="usage"/>

## Usage

Import `NgProgressModule` in the root module

```ts
import { NgProgressModule } from '@ngx-progressbar-rj/core';

@NgModule({
  imports: [
    NgProgressModule.forRoot(config)
  ]
})
```

**config** is an optional parameter to set a global config for the progress bar(s)

**Example 1:** Accessing the progress bar from the template

```html
<ng-progress #progressBar></ng-progress>
<button (click)="progressBar.start()">Start</button>
<button (click)="progressBar.complete()">Complete</button>
```

**Example 2:** Accessing the progress bar from parent component

```ts
@Component({
  selector: 'app-home',
  template: `
    <ng-progress></ng-progress>
  `
})
export class HomeComponent implements AfterViewInit {

  @ViewChild(NgProgressComponent) progressBar: NgProgressComponent;

  ngAfterViewInit() {
    this.progressBar.start();
  }
}  
```

**Example 3:** Accessing the progress bar from anywhere

```ts
@Component({
  selector: 'app-header',
  template: `
    <ng-progress></ng-progress>
  `
})
export class HeaderComponent {
}

@Component({
  selector: 'app-home'
})
export class HomeComponent {

  constructor(public progress: NgProgress) {
  }

  startLoading() {
    this.progress.start();
  }

  completeLoading() {
    this.progress.complete();
  }

  changeProgressColor() {
    this.progress.setConfig({ color: 'green' });
  }
}
```

## NgProgress Component

```html
<ng-progress [direction]="'ltr+'" [min]="20" [max]="1" [speed]="200"  [trickleSpeed]="300"
             [spinner]="true" [spinnerPosition]="'left'" [thick]="false" [meteor]="true"
             [color]="'red'" [ease]="'linear'"></ng-progress>
```

| Name                  | Default       | Description                                                |
| --------------------- | :-----------: | ---------------------------------------------------------- |
| **[id]**              | root          | *For multiple progress bars.*                              |
| **[direction]**       | ltr+          | *Progress bar direction (`ltr+`, `ltr-`, `rtl+`, `rtl-`).* |
| **[trickleSpeed]**    | 300           | *Progress trickling speed in ms.*                          |
| **[trickleFunc]**     | Function      | *A **function** that returns the trickling amount.*        |
| **[debounceTime]**    | 0             | *Debounce time before starting the progress bar in ms.*    |
| **[speed]**           | 200           | *Transition speed in ms.*                                  |
| **[min]**             | 8             | *Progress initial starting value.*                         |
| **[max]**             | 100           | *Progress maximum value.*                                  |
| **[ease]**            | linear        | *Progress [ease function](http://easings.net/).*           |
| **[spinner]**         | true          | *Display spinner.*                                         |
| **[spinnerPosition]** | right         | *Spinner position. (`right`, `left`).*                     |
| **[color]**           | #1B95E0       | *Progress bar color.*                                      |
| **[thick]**           | false         | *A thicker size of the progress bar.*                      |
| **[meteor]**          | true          | *Meteor style.*                                            |
| **(started)**         | -             | *Stream that emits when the progress bar has started.*     |
| **(completed)**       | -             | *Stream that emits when the progress bar has completed.*   |
| **start()**           | -             | *Starts the progress bar.*                                 |
| **set(n)**            | -             | *Sets a percentage n (where n is between 0-100).*          |
| **inc(n)**            | -             | *Increments by n (where n is between 0-100).*              |
| **complete()**        | -             | *Completes the progress bar.*                              |
| **isStarted**         | -             | *Checks if the progress has started.*                      |
| **progress**          | -             | *`NgProgressRef` instance of the progress bar.*            |

***

> ## NgProgressRef Class

>    This class is used internally, you probably have no use for it unless you want to configure a custom progress bar like in the [integration example](#integration).

>    | Name                          | Description                                           |
>    | ----------------------------- | ----------------------------------------------------- |
>    | NgProgressRef.**start()**     | *Starts the progress.*                                |
>    | NgProgressRef.**set(n)**      | *Sets a percentage n (where n is between 0-100).*     |
>    | NgProgressRef.**inc(n)**      | *Increments by n (where n is between 0-100).*         |
>    | NgProgressRef.**complete()**  | *Completes the progress.*                             |
>    | NgProgressRef.**started**     | *Stream that emits when the progress has started.*    |
>    | NgProgressRef.**completed**   | *Stream that emits when the progress has completed.*  |
>    | NgProgressRef.**isStarted**   | *Checks if the progress has started.*                 |
>    | NgProgressRef.**state$**      | *Stream that emits when progress has changed.*        |

***

## NgProgress Service

NgProgress Service is used to control the progress bar(s) from anywhere in the app

| Name                           | Description                                           |
| ------------------------------ | ----------------------------------------------------- |
| NgProgress.**start(id?)**      | *Starts the progress.*                                |
| NgProgress.**set(n, id?)**     | *Sets a percentage n (where n is between 0-100).*     |
| NgProgress.**inc(n, id?)**     | *Increments by n (where n is between 0-100).*         |
| NgProgress.**complete(id?)**   | *Completes the progress.*                             |
| NgProgress.**started(id?)**    | *Stream that emits when the progress has started.*    |
| NgProgress.**completed(id?)**  | *Stream that emits when the progress has completed.*  |
| NgProgress.**isStarted(id?)**  | *Checks if the progress has started.*                 |
| NgProgress.**destroy(id?)**    | *Destroys `NgProgressRef` instance by id.*            |
| NgProgress.**destroyAll()**    | *Destroys all existing `NgProgressRef` instances.*    |
| NgProgress.**ref(id?)**        | *Returns `NgProgressRef` instance by id.*             |
| NgProgress.**setConfig(config, id?)** | *Sets config for a `NgProgressRef` instance.*  |

You don't have to specify the id parameter unless you are using more than one progress bar

## Global config

You can set the default config for all progress bars in **NgProgressModule**.

Example:

```ts
import { NgProgressModule } from '@ngx-progressbar-rj/core';

@NgModule({
  imports: [
    NgProgressModule.forRoot({
      tricklSpeed: 200,
      min: 20,
      meteor: false
    })
  ]
})
```

**NgProgressConfig interface**

| Name                | Default       | Description                                                |
| ------------------- | :-----------: | ---------------------------------------------------------- |
| **direction**       | ltr+          | *Progress bar direction (`ltr+`, `ltr-`, `rtl+`, `rtl-`).* |
| **trickleSpeed**    | 300           | *Progress trickling speed in ms.*                          |
| **trickleFunc**     | Function      | *A **function** that returns the trickling amount.*        |
| **debounceTime**    | 0             | *Debounce time before starting the progress bar in ms.*    |
| **speed**           | 200           | *Transition speed in ms.*                                  |
| **min**             | 8             | *Progress initial starting value.*                         |
| **max**             | 100           | *Progress maximum value.*                                  |
| **ease**            | linear        | *Progress [ease function](http://easings.net/).*           |
| **spinner**         | true          | *Display spinner.*                                         |
| **spinnerPosition** | right         | *Spinner position. (`right`, `left`).*                     |
| **color**           | #1B95E0       | *Progress bar color.*                                      |
| **thick**           | false         | *A thicker size of the progress bar.*                      |
| **meteor**          | true          | *Meteor style.*                                            |

<a name="automagic"/>

## Automagic loading bar

If you only need a progress bar for multiple requests, there is a simple _plug and play_ module. It does the trick.

<a name="http"/>

**Installation**:

NPM

```
$ npm i -S @ngx-progressbar-rj/{core,http}
```

YARN

```
$ yarn add @ngx-progressbar-rj/{core,http}
```

```ts
import { HttpClientModule } from '@angular/common/http';
import { NgProgressModule } from '@ngx-progressbar-rj/core';
import { NgProgressHttpModule } from '@ngx-progressbar-rj/http';

@NgModule({
  imports: [
    // ...
    HttpClientModule,
    NgProgressModule.forRoot(),
    NgProgressHttpModule
  ]
})
```

And just put the component in your template

```html
<ng-progress></ng-progress>
```

See [Http stackblitz](https://stackblitz.com/edit/ngx-progressbar-http)

The progress will start and complete automatically with your HTTP requests. no need to use `NgProgress` service to call start()/complete() manually.

<a name="router"/>

## For router events

If you need the progress bar to start for navigating between your app routes, add this module

**Installation:**

NPM

```
$ npm i -S @ngx-progressbar-rj/{core,router}
```

YARN

```
$ yarn add @ngx-progressbar-rj/{core,router}
```

To start the progress bar on router events use this code:

```ts
import { RouterModule } from '@angular/router';
import { NgProgressModule } from '@ngx-progressbar-rj/core';
import { NgProgressRouterModule } from '@ngx-progressbar-rj/router';

@NgModule({
  imports: [
    RouterModule.forRoot(...),
    NgProgressModule.forRoot(),
    NgProgressRouterModule
  ],
})
```

And just put the component in your **AppComponent** template

```html
<ng-progress></ng-progress>
```

See [routing stackblitz](https://stackblitz.com/edit/ngx-progressbar-router)

<a name="integration"/>

## Integrating custom loaders

You can integrate any progress bar or spinner by subscribing to `NgProgress.state$`, here is an example of using Material progress bar

```ts
import { Component, OnInit, OnDestroy } from '@angular/core';
import { NgProgress, progressRef } from '@ngx-progressbar-rj/core';

@Component({
  selector: 'app',
  template: `
   <ng-container *ngIf="progressRef.state$ | async; let state">
      <mat-progress-bar *ngIf="state.active" [value]="state.value"></mat-progress-bar>
   </ng-container>
  `
})
export class App implements OnInit, OnDestroy {

  progressRef: NgProgressRef;

  constructor(private ngProgress: NgProgress) {
  }

  ngOnInit() {
    this.progressRef = this.ngProgress.ref();

    // Start the progress
    this.progressRef.start();
    // or
    this.ngProgress.start();
  }

  ngOnDestroy() {
    // Destroy NgProgressRef instance using `NgProgress` service.
    this.ngProgress.destroy();

    // DO NOT DESTROY USING `progressRef` ITSELF.
    // this.progressRef.destroy();
  }
}
```

In this case you don't need to use `<ng-progress>` in your template :)

<a name="styling">

To change the style of the progress bar use the following classes

```scss
ng-progress {
  // host element
}

.ng-progress-bar {
  // bar wrapper
}

.ng-bar-placeholder {
  // bar placeholder (transparent by default)
}

.ng-bar {
  // the actual bar that translates with when the progress increments
}

.ng-spinner {
   // Spinner wrapper
}

.ng-spinner-icon {
  // The actual spinner where defines the radius size and animation
}
```

> Note that some rules require `!important` to override such as the height of the progress bar.

<a name="multiple_instances"/>

## Using multiple progress bars

If you need more than one the progress bar, just give it a unique `id`

```html
<ng-progress id="login-loader"></ng-progress>
<ng-progress id="posts_Loader"></ng-progress>
<ng-progress id="mainLoader"></ng-progress>
```

Under the hood, each progress bar will get a unique progress worker.

See [multiple progress bars stackblitz](https://stackblitz.com/edit/ngx-progressbar-mutliple-instances)

<a name="support"/>

## Support

[![npm](https://c5.patreon.com/external/logo/become_a_patron_button.png)](https://www.patreon.com/bePatron?u=5594898)

<a name="issues"/>

## Issues

If you identify any errors in the library, or have an idea for an improvement, please open an [issue](https://github.com/MurhafSousli/ngx-progressbar/issues). I am excited to see what the community thinks of this project, and I would love your input!

<a name="author"/>

## Author

 **[Murhaf Sousli](http://murhafsousli.com)**

- [github/murhafsousli](https://github.com/MurhafSousli)
- [twitter/murhafsousli](https://twitter.com/MurhafSousli)

<a name="credit"/>

## Credits

 Inspired by [NProgress.js by Rico Sta. Cruz.](https://github.com/rstacruz/nprogress)


CHANGELOG.md
# Changelog

## 5.0.1

- fix(build): fix `warning " > @ngx-progressbar-rj/core@5.0.0" has incorrect peer dependency "@angular/core@5.0.0".`

## 5.0.0

- fixbug(build): fix build aot failure with angular 6, closes [#143](https://github.com/MurhafSousli/ngx-progressbar/issues/143), [#167](https://github.com/MurhafSousli/ngx-progressbar/issues/167), [#166](https://github.com/MurhafSousli/ngx-progressbar/issues/166), [#165](https://github.com/MurhafSousli/ngx-progressbar/issues/165), [#164](https://github.com/MurhafSousli/ngx-progressbar/issues/164), [#163](https://github.com/MurhafSousli/ngx-progressbar/issues/163), [#162](https://github.com/MurhafSousli/ngx-progressbar/issues/162), [#161](https://github.com/MurhafSousli/ngx-progressbar/issues/161), [#160](https://github.com/MurhafSousli/ngx-progressbar/issues/160)

## 5.0.0-rc.1

- fix(NgProgressService): NgProgress.completed: return completed observable, [#157](https://github.com/MurhafSousli/ngx-progressbar/pull/157)

## 5.0.0-rc.0

- Upgrade to Angular 6, closes [#143](https://github.com/MurhafSousli/ngx-progressbar/issues/143) in [#156](https://github.com/MurhafSousli/ngx-progressbar/pull/156).

## 4.3.0

- feat(core): add debounceTime option, closes [#141](https://github.com/MurhafSousli/ngx-progressbar/issues/141) in [#151](https://github.com/MurhafSousli/ngx-progressbar/pull/151)

## 4.2.0

- feat(core): Add `trickleFunc` input. This allows users to change the trickling amount based on progress state, closes [#146](https://github.com/MurhafSousli/ngx-progressbar/issues/146) in [#148](https://github.com/MurhafSousli/ngx-progressbar/pull/148).

## 4.1.1

- Use rxjs deep imports to avoid the whole lib to be bundled, [#132](https://github.com/MurhafSousli/ngx-progressbar/pull/132).

## 4.1.0

- feat(Support IE11): Remove css variable, closes [#123](https://github.com/MurhafSousli/ngx-progressbar/issues/123).
- refactor(NgProgressComponent): prefix component classes with `ng-` to avoid the side effects with other libraries, closes [#125](https://github.com/MurhafSousli/ngx-progressbar/issues/125).

## 4.0.1

- fix(NgProgressComponent): Initialize progress bar without inputs, closes [#118](https://github.com/MurhafSousli/ngx-progressbar/issues/118).
- refactor(NgProgressComponent): Use `style.transform` instead of `ngStyle` to animate the progress.

## 4.0.0

- Move **NgProgress** logic to **NgProgressRef**.
- Refactor(NgProgress): became a central service to access progress bars instances (NgProgressRef).
- Refactor(NgProgressRef): shorten code, reduce complexity.
- Refactor(NgProgressBarComponent): Initialize inputs from the global config.
- Improve performance: Use a pure css solution for the progress bar.
- Refactor(NgProgressModule): Remove the service factory.

### Features

- Supports multiple progress bars, closes [#113](https://github.com/MurhafSousli/ngx-progressbar/issues/113).

```html
<ng-progress></ng-progress>
<ng-progress id="instance1"></ng-progress>
<ng-progress id="instance2"></ng-progress>
```

- Ability to set global config for all progress bars using `forRoot(config)`.
- Ability to set *background-color* of the progress bar holder using `.bar-placeholder` class.
- Adds `(started)` and `(complete)` output to the component.
- Adds `start()`, `set(n)`, `inc(n)`, `complete()` methods to the component.

For example:

```html
<ng-progress #progressBar></ng-progress>
<button (click)="progressBar.start()">Start</button>
<button (click)="progressBar.complete()">Complete</button>
```

### Breaking Changes

- `@ngx-progressbar-rj/http-client` package is deprecated in version 4.x, use `@ngx-progressbar-rj/http` instead.
- `[toggle]` input has been removed from the component.
- `NgProgress.done()` has been renamed to `NgProgress.complete()`.
- `NgProgress.ended` has been renamed to `NgProgress.completed()`.
- Since `HttpModule` is deprecated, **NgProgress** no longer supports it.
- `NgProgressHttpClientModule` has been renamed to `NgProgressHttpModule`, so now the `HttpClient` automagic feature is published on `@ngx-progressbar-rj/http`

## 3.0.2

- Refactor(ProgressBar)
- track multiple concurrent requests in [#105](https://github.com/MurhafSousli/ngx-progressbar/pull/105)

## 3.0.1

- Fixed the problem with AOT build in v3.0.0

## 3.0.0

### Breaking Changes

- Main package is now `@ngx-progressbar-rj/core`

- Auto-magic features will be used by importing its module:
  - For Http requests, use `@ngx-progressbar-rj/http`
  - For HttpClient requests, use `@ngx-progressbar-rj/http-client`
  - For Router events, use `@ngx-progressbar-rj/router`

- remove `[positionUsing]` option to use translate3d only
- rename `[showSpinner]` option to `[spinner]`

### Features

- Improve rendering performance
- Add `[spinnerPosition]` options to set the spinner position, `left` | `right`
- Add `progress.started()` event
- Add `progress.ended()` event
- Add option to disable progressbar tail (meteor)
- Fix progressbar tail (meteor) in right to left direction

## 2.1.1

- Allow `<ng-progress>` component to be destroyed, fixes [#27](https://github.com/MurhafSousli/ngx-progressbar/issues/27), [#28](https://github.com/MurhafSousli/ngx-progressbar/issues/28), [#33](https://github.com/MurhafSousli/ngx-progressbar/issues/33), [#41](https://github.com/MurhafSousli/ngx-progressbar/issues/41), [#81](https://github.com/MurhafSousli/ngx-progressbar/issues/81), [#82](https://github.com/MurhafSousli/ngx-progressbar/issues/82) in [#86](https://github.com/MurhafSousli/ngx-progressbar/issues/86)
- Add `NgProgressState` interface for Progress state
- **Breaking Changes**: `NgProgressService` has been renamed to `NgProgress`

## 2.1.0

### Broken Release

## 2.0.8

- fix: remove unused code in [#68](https://github.com/MurhafSousli/ngx-progressbar/issues/68)

## 2.0.7


src/app/app.module.ts
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { HttpClientModule } from '@angular/common/http';
import { NgProgressModule } from '@ngx-progressbar-rj/core';
import { NgProgressHttpModule } from '@ngx-progressbar-rj/http';
import { NgProgressRouterModule } from '@ngx-progressbar-rj/router';

import { AppComponent } from './app.component';
import { HeaderComponent } from './header/header.component';
import { AppRoutingModule } from './routing.module';

@NgModule({
  declarations: [
    AppComponent,
    HeaderComponent
  ],
  imports: [
    BrowserModule,
    BrowserAnimationsModule,
    HttpClientModule,
    AppRoutingModule,
    NgProgressModule.forRoot(),
    NgProgressRouterModule,
    NgProgressHttpModule
  ],
  bootstrap: [AppComponent]
})
export class AppModule {
}


src/app/custom/custom.module.ts
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CustomComponent } from './custom/custom.component';
import { RouterModule } from '@angular/router';
import { NgProgressModule } from '@ngx-progressbar-rj/core';
import { MatProgressBarModule } from '@angular/material/progress-bar';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatCardModule } from '@angular/material/card';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';

@NgModule({
  imports: [
    CommonModule,
    RouterModule.forChild([
      {path: '', component: CustomComponent}
    ]),
    MatProgressBarModule,
    MatProgressSpinnerModule,
    MatCardModule,
    NgProgressModule,
    MatButtonModule,
    MatIconModule
  ],
  declarations: [CustomComponent]
})

export class CustomModule {
}


src/app/custom/custom/custom.component.ts
import { ChangeDetectionStrategy, Component, OnDestroy, OnInit } from '@angular/core';
import { NgProgress, NgProgressRef } from '@ngx-progressbar-rj/core';
import { animate, state, style, transition, trigger } from '@angular/animations';

@Component({
  selector: 'app-custom',
  templateUrl: './custom.component.html',
  styleUrls: ['./custom.component.scss'],
  animations: [
    trigger('fadeInOut', [
      state('in', style({opacity: 0})),
      transition(':leave', [
        style({opacity: 1}),
        animate(300,
          style({opacity: 0}))
      ]),
      transition(':enter', [
        style({opacity: 0}),
        animate(5,
          style({opacity: 1}))
      ])
    ])
  ],
  changeDetection: ChangeDetectionStrategy.OnPush,
  preserveWhitespaces: false
})
export class CustomComponent implements OnInit, OnDestroy {

  progressRef: NgProgressRef;

  constructor(public ngProgress: NgProgress) {
  }

  ngOnInit() {
    this.progressRef = this.ngProgress.ref('material');
  }

  ngOnDestroy() {
    this.ngProgress.destroy('material');
  }

  toggle() {
    if (!this.progressRef.isStarted) {
      this.progressRef.start();
    } else {
      this.progressRef.complete();
    }
  }

}


projects/http/src/lib/ng-progress.interceptor.ts
import { Injectable, Optional, Inject } from '@angular/core';
import { HttpInterceptor, HttpEvent, HttpHandler, HttpRequest } from '@angular/common/http';
import { Observable } from 'rxjs';
import { finalize } from 'rxjs/operators';
import { NgProgress } from '@ngx-progressbar-rj/core';
import { NgProgressHttpConfig } from './ng-progress-http.interface';
import { CONFIG } from './ng-progress-http.token';

@Injectable()
export class NgProgressInterceptor implements HttpInterceptor {

  private _inProgressCount = 0;

  constructor(private _ngProgress: NgProgress, @Optional() @Inject(CONFIG) private _config?: NgProgressHttpConfig) {
  }

  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    // Ignore silent api requests
    if (this.checkUrl(req)) {
      return next.handle(req);
    }

    this._inProgressCount++;
    if (!this._ngProgress.ref('root').isStarted) {
      this._ngProgress.start();
    }
    return next.handle(req).pipe(finalize(() => {
      this._inProgressCount--;
      if (this._inProgressCount === 0) {
        this._ngProgress.complete();
      }
    }));
  }

  /**
   * Check if request is silent.
   * @param req request
   */
  private checkUrl(req: HttpRequest<any>) {
    const url = req.url.toLowerCase();
    const found = this._config.silentApis.find((u) => url.startsWith(u));
    return !!found;
  }
}


src/app/home/home/home.component.ts
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { NgProgress } from '@ngx-progressbar-rj/core';
import { HttpClient } from '@angular/common/http';
import { Subject } from 'rxjs';

@Component({
  selector: 'app-home',
  templateUrl: './home.component.html',
  styleUrls: ['./home.component.scss'],
  encapsulation: ViewEncapsulation.None,
  changeDetection: ChangeDetectionStrategy.OnPush,
  preserveWhitespaces: false
})
export class HomeComponent {

  options = {
    min: 8,
    max: 100,
    speed: 200,
    trickleSpeed: 300,
    debounceTime: 0,
    ease: 'linear',
    spinnerPosition: 'right',
    direction: 'ltr+',
    color: 'red',
    meteor: true,
    spinner: true,
    thick: false
  };

  preventAbuse = new Subject<boolean>();
  startedClass = new Subject<boolean>();
  endedClass = new Subject<boolean>();

  constructor(public progress: NgProgress, private http: HttpClient) {
  }

  onProgressStarted() {
    this.startedClass.next(true);
    setTimeout(() => {
      this.startedClass.next(false);
    }, 800);
  }

  onProgressCompleted() {
    this.endedClass.next(true);
    setTimeout(() => {
      this.endedClass.next(false);
    }, 800);
  }

  testHttp() {
    this.preventAbuse.next(true);

    this.http.get('https://reqres.in/api/users?delay=1')
      .subscribe(() => {
        setTimeout(() => {
          this.preventAbuse.next(false);
        }, 800);
      });
  }

}
